### PR TITLE
Feature/#40 add moshi type adapters

### LINKED ISSUE #40: Add moshi type adapters from app

### PR DESCRIPTION
**Pull request by biafra23**
Closes #40  .

Changes proposed in this pull request:
- Adds JSON adapters for addresses and Wei values
- Creates a default builder that can be modified by clients if they need additional TypeAdapters
- String.parseToBigInteger() has been added to NumberUtils and can be removed from StringUtils in the app 

At the moment these Adapters are not used in the gnosis-safe-android app. And the legacy app does not support the changes made in v0.11.0 

@gnosis/mobile-devs

### CODE REVIEW

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package pm.gnosis.eip712.adapters.moshi
```
Comment (rmeissner) [inline eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:1]: Should we really put them in the eip712 module? They are not really related, right? (EIP-712 is about signing typed data and these type adapter are more general)

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+class DecimalNumberAdapter {
+    @ToJson
+    fun toJson(@DecimalNumber bigInteger: BigInteger): String = bigInteger.toString()
+
+    @FromJson
+    @DecimalNumber
+    fun fromJson(decimalNumber: String): BigInteger = decimalNumber.toBigInteger()
+}
+
+class DefaultNumberAdapter {
```
Comment (rmeissner) [inline eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:53]: I would rename this one to HexNunberAdapter

Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:53]: Ohhh we have that already, maxbe then this one should try to detect if it is a decimal or hex number based on the prefix

Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:53]: But not sure how to handle the "toJson" in that case

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            .add(DecimalNumberAdapter())
+            .add(DefaultNumberAdapter())
+            .add(SolidityAddressAdapter())
+    }
+}
+
+class WeiAdapter {
+    @ToJson
+    fun toJson(wei: Wei): String =
+        StringBuilder("0x").append(wei.value.toString(16)).toString()
```
Comment (rmeissner) [inline eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:27]: This should be the same as `wei.value.toHexStrin()`

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 2 above)
Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:53]: Maybe toJson should default to the hex representation (it is the default for most ethereum related logic)

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 1 above)
Comment (biafra23) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:1]: I agree. Is there any other module that's close enough? Or should this be in a separate module?

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 3 above)
Comment (biafra23) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:27]: I had the same thought. Was just to slow to commit it already...

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 1 above)
Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:1]: Good question ... I thought about the utils module but that would mean we add moshi as a dependency there ... Maybe a separate one is cleaner .... Not sure to be honest

Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:1]: Android-common also feels wrong as moshi is not only for android

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 2 above)
Comment (biafra23) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:56]: Then we can remove it. But I think this is supposed to be the fallback if the passed number is not annotated as @HexNumber.

Comment (rmeissner) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:56]: Yes saw this too, as said maybe keep it and use the parseToBigInteger function in the fromJson part

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 1 above)
Comment (jpalvarezl) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt:1]: My suggestion would be to maybe add an additional library that follows the nameing of the `:models` module. Something like `:models-moshi`

Comment (biafra23) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt]: I called it common-adapter-moshi in 7eae9130b0e4cf87825450a5013e8370d8c0c160

File: eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt (hunk 2 above)
Comment (biafra23) [reply eip712-adapter-moshi/src/main/java/pm/gnosis/eip712/adapters/moshi/TypeAdapters.kt]: Fixed in 6368c01e9b0d89219f2cfef54a7d13dc3a62e2c8

File: build.gradle [hunk 4]
```diff
@@ -4,6 +4,9 @@ buildscript {
     apply from: 'buildsystem/constants.gradle'
     apply from: 'buildsystem/helpers.gradle'
     apply from: 'buildsystem/dependencies.gradle'
+    ext {
+        versions.kotlin = '1.3.61'
```
Comment (jpalvarezl) [inline build.gradle:8]: This one is not necessary, as it the value is already available in `buildsystem/dependencies.gradle` and is actually pointing to a newer release.

File: common-adapter-moshi/build.gradle [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+dependencies {
+    // We specify junit before anything else to make sure that this version is prioritized over the
+    // one bundled in the android.jar
+    testImplementation "junit:junit:$versions.junit"
+
+    implementation project(":models")
+    implementation project(":utils")
+    implementation project(":crypto")
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib"
```
Comment (jpalvarezl) [inline common-adapter-moshi/build.gradle:18]: This should probably look like `implementation "org.jetbrains.kotlin:kotlin-stdlib:$versions.kotlin"`

File: common-adapter-moshi/proguard-rules.pro [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (jpalvarezl) [inline common-adapter-moshi/proguard-rules.pro:1]: There is no need for this file nor the `consumer-rules.pro` above☝️ . Proguard rules should probably be defined in the app consuming these libraries.

File: TypeAdapters.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
+package pm.gnosis.eip712.adapters.moshi
```
Comment (jpalvarezl) [inline TypeAdapters.kt:1]: So this file should probably be removed from this module then?

Comment (biafra23) [reply TypeAdapters.kt:1]: Ouch. Yes.

File: build.gradle (hunk 4 above)
Comment (biafra23) [reply build.gradle]: Fixed in cb30a2d8773d8303c1fb68fcd7dc1bd13b0571ce

File: common-adapter-moshi/proguard-rules.pro (hunk 6 above)
Comment (biafra23) [reply common-adapter-moshi/proguard-rules.pro]: Fixed in cb30a2d8773d8303c1fb68fcd7dc1bd13b0571ce

File: TypeAdapters.kt (hunk 7 above)
Comment (biafra23) [reply TypeAdapters.kt]: Fixed in a63f0af36cafc2fa3861eaa5d0e24dacb35ffc31

**Comment by jpalvarezl** [comment]
LGTM

**Pull request merged**